### ISSUE TITLE
Profile multi-minute CPU spins

### ISSUE THREAD

**Issue by Kegsay** [matrix-org/matrix-appservice-irc#404]
Occasionally I'm seeing 4-5min periods of 100% CPU spin on Freenode. Messages both ways get delayed for the duration of the spin. We need to [profile](https://github.com/node-inspector/v8-profiler) why this is happening. Stuff is still being logged, but the logs themselves don't show anything interesting.

**Comment by Kegsay** [comment]
![freenode-cpu](https://cloud.githubusercontent.com/assets/7190048/24302281/053dc8c4-10aa-11e7-8475-b6551b19d385.png)

**Comment by Kegsay** [comment]
This is really strange. The event loop can't be getting overloaded because:
 - Sending a message on IRC during the spin is logged almost instantly, meaning we were able to receive the bytes down the wire from IRC without any trouble.
 - No IRC connections ping timeout during the spin.
 - Delayed timers fire near enough on the 10 second mark (meaning `setTimeout` isn't being delayed).
 - Sending a message on Matrix during the spin is logged almost instantly, meaning we were able to receive the HTTP request *and reply to it*.

I don't know where the CPU time is being sunk into though. It must be a path which affects both I->M and M->I as both ends get delayed **after** they log the message. This is smelling more and more like it's the database.

**Comment by Kegsay** [comment]
I've done the obvious optimisation on the I->M user id mapping path, but I want to optimise room sending both ways (so we don't actually yield between getting the message and sending it through in the general case). This requires a cache of `room_id > [#chan] ` and `#chan > [room_id] ` depending on the direction. 

The problem here is when to invalidate the cache: we support provisioned mappings so we need to be sure our cache is kept up to date when people add or remove a mapping. This should be doable though given they all go through the same data structures in the end.

**Comment by Kegsay** [comment]
I've attached CPU profiling to USR2 and will keep this issue updated with latest findings. So far, it's shown:
 - Cumulative `Array.isArray` checks is expensive.
 - This is being done by `_roomStore.getEntriesByLinkData`.
 - This is called by `DataStore.getMappingsForChannelByOrigin` (called in `IrcHandler.onTopic`) and `DataStore.getRoomIdsFromConfig` (called once at startup)

tl;dr - receiving topics from IRC is expensive (notably so on startup!!). We do this check in order to:
```js
// Only bridge topics for rooms created by the bridge, via !join or an alias
```

The reason why it's SO expensive is because we fetch ALL THE ENTRIE evar:
```js
this._roomStore.getEntriesByLinkData({})
```

we can surely scope this better given we know the channel.

**Comment by Kegsay** [comment]
It also looks like my suspicions about nedb being slow are correct - the sheer bulk of time (41.6%) is spent in a single NEDB function: `deepCopy`. Somewhat surprisingly, 24.8% of time is spent doing GC.

**Comment by Kegsay** [comment]
Right, now I know why it spins at 100% CPU and yet doesn't seem to cause ping timeouts IRC-side. All the work is being done in a `setImmediate` block which:

> Schedules the "immediate" execution of the callback after I/O events' callbacks and before timers created using setTimeout() and setInterval() are triggered. Returns an Immediate for use with clearImmediate().
>
> When multiple calls to setImmediate() are made, the callback functions are queued for execution in the order in which they are created. The entire callback queue is processed every event loop iteration. If an immediate timer is queued from inside an executing callback, that timer will not be triggered until the next event loop iteration.

(better described http://stackoverflow.com/a/20296580 )

**Comment by Kegsay** [comment]
So there seems to be 2 actionable things we can do, neither of which are related to 100% cpu spin *after* starting up:
 - Make I->M topics cheaper to execute at startup - we get 1 topic for every channel at startup, causing 1 fetch-the-world query which can take up to 1s to deep copy everything(!) - we have around 4k channels so that's tens of minutes being wasted. See https://github.com/matrix-org/matrix-appservice-irc/issues/404#issuecomment-289728574
 - Share `Intent` data structures in `matrix-appservice-bridge` : We were going to do this for memory reasons anyway, but it's also somewhat telling at how badly this scales (looping all created Matrix user `Intents` for each incoming event). Some work on this has already been made. https://github.com/matrix-org/matrix-appservice-bridge/tree/kegan/share-intent-data

When I catch the Freenode bridge in the act of spinning now, I'll hopefully be able to add a 3rd entry and then that should be everything.

**Comment by Kegsay** [comment]
Caught it spinning, and the main culprit is nedb's `match` function which took 38.8% total time. Same problem with `setImmediate` starving the event loop. I still don't know *why* it starts to spin, but at least I now know *what* it is actually doing with the cpu time. It looks like doing:

> a cache of room_id > [#chan] and #chan > [room_id] depending on the direction.

will help greatly, since we'll be able to relay messages in-between `processImmediate` calls if we do it synchronously. This is an unsatisfying fix though, as I still don't know what the high-level database call is that making it do this expensive `match`. It's worth noting however that just having all the data be in-memory doesn't help here, it just seems to be spending its life trying to match against a large dataset.

**Comment by Kegsay** [comment]
So I ran some benchmarks on the actual machine which runs the Freenode bridge which tests each and every `DataStore` function with realistic data to get a feel for how long things can take. This turned up a few surprises. The slow ones (>100ms) are below:
```
Avg=776.3ms Max= 1585ms Min=  683ms Total=10  getMatrixUserByUsername - called when generating an ident username for a new connection
Avg=141.9ms Max=  146ms Min=  137ms Total=10  getRoomIdsFromConfig - called on startup
Avg=372.4ms Max=  397ms Min=  363ms Total=10  getTrackedChannelsForServer - called on startup
Avg=  377ms Max=  402ms Min=  371ms Total=10  getAllChannelMappings - called when a channel gets +s, called when initial member list syncing
Avg=376.2ms Max=  947ms Min=  263ms Total=10  getMappingsForChannelByOrigin - called for every I->M topic
```

`getMatrixUserByUsername` is a surprise.

**Comment by Kegsay** [comment]
So, I've finally found the cause of the "random" CPU spins. In a nutshell, it's caused by brand-new 3rd-party bridged users joining a freenode channel. After I posted the above comment, I mulled my benchmark results over lunch (why do all epiphanies come when you're not at the keyboard?). In particular, I was mulling over the surprise result: `getMatrixUserByUsername` is very slow. It's called whenever we generate a brand-new ident username in order to avoid getting clashes and assigning the same username for multiple Matrix users. The way that code is implemented means it will call that for every prospective username, and add 1 to the result if there was a clash, here's a snippet of what it does:
```js
    while (true) {
        let usr = yield this.dataStore.getMatrixUserByUsername(domain, uname);
        if (usr && usr.getId() !== userId) { // occupied username!
            if (!modifyUsername()) {
                throw new Error("Ran out of entries: " + uname);
            }
        }
        else {
            if (!usr) {
                log.info(
                    "Generated ident username %s for %s on %s",
                    uname, userId, domain
                );
            }
            else {
                log.info(
                    "Returning cached ident username %s for %s on %s",
                    uname, userId, domain
                );
            }
            break;
        }
    }
```

My thought was: "What if this code is going berserk and is calling `getMatrixUserByUsername` multiple times? Why would it do that?" Well, it does that for clashes, but we only clash if users have a very long shared prefix on the `user_id`. We log when we have generated the username, so I went back through previous cpu spins to see if this may have been the cause. Sure enough:
 - CPU spin between 10:38 ~ 10:43:
    ```
    2017-03-29 10:43:28 INFO:IdentGenerator Generated ident username telegr_265 for @telegram_276561634:tchncs.de on chat.freenode.net
    ```
 - CPU spin between 13:08 ~ 13:18:
    ```
    2017-03-29 13:12:44 INFO:IdentGenerator Generated ident username telegr_266 for @telegram_351857209:tchncs.de on chat.freenode.net
    2017-03-29 13:18:29 INFO:IdentGenerator Generated ident username telegr_267 for @telegram_302189448:tchncs.de on chat.freenode.net
    ```
 - CPU spin between 05:48 ~ 05:53:
    ```
    2017-03-29 05:53:14 INFO:IdentGenerator Generated ident username telegr_264 for @telegram_257110580:tchncs.de on chat.freenode.net
    ```
 - CPU spin between 04:12 ~ 04:15:
    ```
    2017-03-29 04:14:45 INFO:IdentGenerator Generated ident username slackt_135 for @_slack_tendermint_U4RT41T63:matrix.org on chat.freenode.net
    ```

Back of the envelope calculations for the latency induced by calling `getMatrixUserByUsername` 260 times = 260 * 750ms = 3.25mins, so it all roughly adds up.

So why is `getMatrixUserByUsername` so slow? It's pretty simple really, we don't ever index that field. Adding the index should fix this.

**Comment by Kegsay** [comment]
Indexing fixes it. Pre-index:
```
Avg=958.1ms Max= 1778ms Min=  852ms Total=10  getMatrixUserByUsername
{"rss":655560704,"heapTotal":621748224,"heapUsed":539775504}
```

Post-indexing:
```
Avg=  0.8ms Max=    5ms Min=    0ms Total=10  getMatrixUserByUsername
{"rss":1205485568,"heapTotal":1106284544,"heapUsed":1070784344}
```

Nasty memory hit though (but this is freenode's user database).

**Comment by Kegsay** [comment]
https://github.com/matrix-org/matrix-appservice-irc/commit/8e3065efdf7a05c6653c8e0f549e009d35fdcb24 should help speed up reconnections on startup, given most of the CPU time was going into that operation.

**Comment by Kegsay** [comment]
Now working on `PublicitySyncer._solveVisibility`, which is slow on startup as it works out if a room should be visible or not in the directory, which also uses `getAllChannelMappings`. Here's the before stats:
```
Avg=6311.7ms Max= 6490ms Min= 6259ms Total=10  _solveVisibility
Avg=516.9ms Max= 1534ms Min=  376ms Total=10  getAllChannelMappings
```

It looks like we have an O(N^2) algorithm here, because we don't store the reverse mapping of `#channel` to `!room:id`, so for each channel we loop every single room in a map and check the channels array. Keeping a reverse mapping cuts it down to:

```
Avg=393.9ms Max=  413ms Min=  389ms Total=10  _solveVisibility
Avg=351.7ms Max=  356ms Min=  349ms Total=10  getAllChannelMappings
```

making the bulk of the time being spent in `getAllChannelMappings`. This coupled with the previous commit should make startup times a LOT faster cpu wise.

**Comment by Kegsay** [comment]
This landed and is running on Freenode and is performing *wonderfully*. Closing.
